x-pack/plugin/security/src/main/plugin-metadata/plugin-security.policy

```diff
@@ -31,6 +31,10 @@ grant {
   // needed for kerberos file permission tests to access user information
   permission java.lang.RuntimePermission "accessUserInformation";
   permission java.lang.RuntimePermission "getFileStoreAttributes";
+
+  // needed for OIDC and JWT realm nimbus-jose-jwt dependency
+  permission java.lang.RuntimePermission "accessDeclaredMembers";
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```

Comment:
The reason this is here is that when serializing classes [without a `TypeAdapter`](https://github.com/google/gson/blob/main/gson/src/main/java/com/google/gson/Gson.java#L294), Gson falls back on modifying field visibility for serialization. One such class is `Nonce`. When constructing a `SignedJwtToken` instance with a Nonce claim like we do here:

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticatorTests.java#L313-L333

we get stacktraces like:

```
Failed making field 'com.nimbusds.oauth2.sdk.id.Identifier#value' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
com.nimbusds.jose.shaded.gson.JsonIOException: Failed making field 'com.nimbusds.oauth2.sdk.id.Identifier#value' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
	at __randomizedtesting.SeedInfo.seed([F4B5567549108E6:97D65E57480DDB41]:0)
	at app//com.nimbusds.jose.shaded.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:38)
	at app//com.nimbusds.jose.shaded.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:287)
	at app//com.nimbusds.jose.shaded.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
	at app//com.nimbusds.jose.shaded.gson.Gson.getAdapter(Gson.java:546)
	at app//com.nimbusds.jose.shaded.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:55)
	at app//com.nimbusds.jose.shaded.gson.internal.bind.MapTypeAdapterFactory$Adapter.write(MapTypeAdapterFactory.java:207)
	at app//com.nimbusds.jose.shaded.gson.internal.bind.MapTypeAdapterFactory$Adapter.write(MapTypeAdapterFactory.java:144)
	at app//com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:825)
	at app//com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:795)
	at app//com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:742)
	at app//com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:719)
	at app//com.nimbusds.jose.util.JSONObjectUtils.toJSONString(JSONObjectUtils.java:527)
	at app//com.nimbusds.jose.Payload.toString(Payload.java:388)
	at app//com.nimbusds.jose.Payload.toBytes(Payload.java:420)
	at app//com.nimbusds.jose.Payload.toBase64URL(Payload.java:437)
	at app//com.nimbusds.jose.JWSObject.composeSigningInput(JWSObject.java:209)
	at app//com.nimbusds.jose.JWSObject.<init>(JWSObject.java:119)
	at app//com.nimbusds.jwt.SignedJWT.<init>(SignedJWT.java:61)
	at app//org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectTestCase.generateIdToken(OpenIdConnectTestCase.java:88)
	at app//org.elasticsearch.xpack.security.authc.oidc.OpenIdConnectRealmTests.testBuildLogoutResponseFromEndsessionEndpointWithExistingParameters(OpenIdConnectRealmTests.java:382)
	at java.base@20-ea/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base@20-ea/java.lang.reflect.Method.invoke(Method.java:578)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
	at app//org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at app//org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at app//org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at app//org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
	at app//com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
	at app//com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
	at app//com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
	at app//org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
	at app//com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at app//com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at app//org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at app//org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at app//org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at app//org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
	at app//com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at app//com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
	at app//com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
	at java.base@20-ea/java.lang.Thread.run(Thread.java:1623)
Caused by: java.security.AccessControlException: access denied ("java.lang.reflect.ReflectPermission" "suppressAccessChecks")
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:488)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:1071)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
	at java.base/java.lang.reflect.AccessibleObject.checkPermission(AccessibleObject.java:91)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:174)
	at com.nimbusds.jose.shaded.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:35)
	... 55 more
```

For ES, this only ever happens in unit test code and only for `claim("nonce", nonce)` which we _could_ work around by converting the nonce instance to a String. However, `jose-jwt` is also used by another third party dependency `oauth2-oidc-sdk`, so I'm opting for robustness here and granting the requisite permission. Admittedly, I haven't come up with a code path where we _would_ actually get a claim that present a problem (generally we should only get claims of the "standard" types like `String` etc), but again, being cautious here.

Reply:
Mostly curious: How hard is it to register a custom TypeAdapter for `Nonce`?

Reply:
It's actually quite easy, it turns out. Just need to add:

```java
static {
  JSONValue.registerWriter(Nonce.class, new JsonWriterI<Nonce>() {
      @Override
      public <E extends Nonce> void writeJSONString(E e, Appendable appendable, JSONStyle jsonStyle) throws IOException {
          appendable.append(e.toJSONString());
      }
  });
}
``` 

to, e.g., `OpenIdConnectAuthenticator` -- so that's actually a nice alternative

Reply:
ha! This is the first time that I've seen "suppressAccessChecks" - one of the most powerful permissions in the JDK. The holder of this permission can use it to effectively bypass the security manger and do anything :-)

Reply:
BTW - OI'm ok with this. We grant it elsewhere too - not great, but it is what it is!